src/Servers/HttpSys/src/MessagePump.cs

```diff
@@ -121,11 +119,11 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
             }
 
             // Can't call Start twice
-            Contract.Assert(_application == null);
+            Contract.Assert(_getRequestWorker == null);
 
             Contract.Assert(application != null);
 
-            _application = new ApplicationWrapper<TContext>(application);
+            _getRequestWorker = (server, nativeRequestContext) => ApplicationWrapper<TContext>.RentWrapper(server, nativeRequestContext, this, application);
```

Comment:
@davidfowl has pointed out; cognitively this is probably better as a factory class rather than a capturing lambda (as per the version in IIS)

Reply:
Fxied